### PR TITLE
prep work for enabling caching

### PR DESCRIPTION
**Pull request by makdharma**
Added new header grpc-payload-bin
Added new channel arg for setting max payload size
Ability to create a GET request in client filter
Ability to parse the payload from header in server filter.

### CODE REVIEW

File: src/core/lib/channel/http_client_filter.c [hunk 1]
```diff
@@ -134,17 +136,43 @@ static void hc_mutate_op(grpc_call_element *elem,
...
+      op->send_message->length < channeld->max_payload_size_for_get) {
+    method = GRPC_MDELEM_METHOD_GET;
+  } else if (op->send_initial_metadata_flags &
+             GRPC_INITIAL_METADATA_IDEMPOTENT_REQUEST) {
+    method = GRPC_MDELEM_METHOD_PUT;
+  }
+
+  if (method == GRPC_MDELEM_METHOD_GET) {
+    gpr_slice slice;
+    /* TODO (makdharma): extend code for messages with multiple slices */
```
Comment (ctiller) [inline src/core/lib/channel/http_client_filter.c]: Needs to be completed before submission

File: src/core/lib/transport/static_metadata.c [hunk 2]
```diff
@@ -1,11 +1,11 @@
 /*
```
Comment (ctiller) [inline src/core/lib/transport/static_metadata.c]: clang-format

**Comment by makdharma** [comment]
test this please

**Comment by makdharma** [comment]
@nicolasnoble Test failures look like a flake, could you please confirm and LGTM if so?

**Comment by nicolasnoble** [comment]
I don't have a clear idea as to why clang-format seems to have lost its head here and is fighting with itself. We've been taking precautions to make sure we are sticking on a very specific version of it for this exact reason.

File: src/core/lib/channel/http_client_filter.c [hunk 3]
```diff
@@ -134,17 +150,57 @@ static void hc_mutate_op(grpc_call_element *elem,
   /* grab pointers to our data from the call element */
   call_data *calld = elem->call_data;
   channel_data *channeld = elem->channel_data;
+
+  /* Decide which HTTP VERB to use */
+  grpc_mdelem *method = GRPC_MDELEM_METHOD_POST;
+  if (op->send_initial_metadata != NULL && (op->send_initial_metadata_flags &
```
Comment (ctiller) [inline src/core/lib/channel/http_client_filter.c]: Comment that we need both send initial metadata and the first send in the same batch (for now) to enable GET

File: src/core/lib/channel/http_client_filter.h [hunk 4]
```diff
@@ -41,4 +41,8 @@ extern const grpc_channel_filter grpc_http_client_filter;
 /* Channel arg to override the http2 :scheme header */
 #define GRPC_ARG_HTTP2_SCHEME "grpc.http2_scheme"
 
+/* Channel arg to determine maximum size of payload eligable for GET request */
+#define GRPC_ARG_MAX_PAYLOAD_SIZE_FOR_GET "grpc.max_payload_size_for_get"
+static const size_t kMaxPayloadSizeForGet = 2048;
```
Comment (ctiller) [inline src/core/lib/channel/http_client_filter.h]: move to implementation file (no need for this to be exposed)

File: src/core/lib/channel/http_client_filter.c [hunk 5]
```diff
@@ -134,17 +150,57 @@ static void hc_mutate_op(grpc_call_element *elem,
...
+    gpr_slice_buffer slices;
+    gpr_slice_buffer_init(&slices);
+    /* allocate memory to hold the entire payload */
+    calld->payload_bytes = gpr_malloc(op->send_message->length);
+    GPR_ASSERT(calld->payload_bytes);
+    uint8_t *wrptr = calld->payload_bytes;
+
+    /* copy payload from slices into payload_bytes. It gets freed in
+     * op_complete*/
+    while (grpc_byte_stream_next(NULL, op->send_message, &slice, ~(size_t)0,
```
Comment (ctiller) [inline src/core/lib/channel/http_client_filter.c]: Bad usage of grpc_byte_stream_next.

```
/* returns 1 if the bytes are available immediately (in which case
 * on_complete will not be called), 0 if the bytes will be available
 * asynchronously.
 *
 * max_size_hint can be set as a hint as to the maximum number
 * of bytes that would be acceptable to read.
 *
 * once a slice is returned into *slice, it is owned by the caller.
 */
int grpc_byte_stream_next(grpc_exec_ctx *exec_ctx,
                          grpc_byte_stream *byte_stream, gpr_slice *slice,
                          size_t max_size_hint, grpc_closure *on_complete);
```

This needs a closure chain to allow asynchronous fetching of the bytes. Not doing so leaves us with a time-bomb.

File: src/core/lib/channel/http_client_filter.h (hunk 4 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.h]: We need to expose this channel arg. It is the only way a client app can prevent GET requests at run-time.

File: src/core/lib/channel/http_client_filter.c [hunk 6]
```diff
@@ -129,22 +166,105 @@ static grpc_mdelem *client_strip_filter(void *user_data, grpc_mdelem *md) {
...
                          grpc_transport_stream_op *op) {
   /* grab pointers to our data from the call element */
   call_data *calld = elem->call_data;
   channel_data *channeld = elem->channel_data;
+
+  /* Decide which HTTP VERB to use. We use GET if the request is marked
+  cacheable, and the operation contains both initial metadata and send message,
+  and the payload is below the size threshold, and all the data
+  for this request is immediately available. */
+  grpc_mdelem *method = GRPC_MDELEM_METHOD_POST;
```
Comment (ctiller) [inline src/core/lib/channel/http_client_filter.c]: This whole block (up until line 238) could probably go into an `if (op->send_initial_metadata)` right?
That way we'd skip doing some complicated processing on other code paths (like, say, recv_initial_metadata)

File: src/core/lib/channel/http_client_filter.c [hunk 7]
```diff
@@ -119,6 +138,24 @@ static void hc_on_recv(grpc_exec_ctx *exec_ctx, void *user_data,
...
+  grpc_call_element *elem = user_data;
+  call_data *calld = elem->call_data;
+  if (calld->payload_bytes) {
+    gpr_free(calld->payload_bytes);
+    calld->payload_bytes = NULL;
+  }
+  calld->on_complete->cb(exec_ctx, calld->on_complete->cb_arg, error);
+}
+
+static void send_done(grpc_exec_ctx *exec_ctx, void *elemp, grpc_error *error) {
```
Comment (ctiller) [inline src/core/lib/channel/http_client_filter.c]: This should just be a call to continue_send_message

File: src/core/lib/channel/http_client_filter.c [hunk 8]
```diff
@@ -129,22 +166,105 @@ static grpc_mdelem *client_strip_filter(void *user_data, grpc_mdelem *md) {
...
+  uint8_t *wrptr = calld->payload_bytes;
+  while (grpc_byte_stream_next(exec_ctx, calld->send_op.send_message,
+                               &calld->incoming_slice, ~(size_t)0,
+                               &calld->got_slice)) {
+    memcpy(wrptr, GPR_SLICE_START_PTR(calld->incoming_slice),
+           GPR_SLICE_LENGTH(calld->incoming_slice));
+    wrptr += GPR_SLICE_LENGTH(calld->incoming_slice);
+    gpr_slice_buffer_add(&calld->slices, calld->incoming_slice);
+    if (calld->send_length == calld->slices.length) {
+      calld->send_message_blocked = false;
```
Comment (ctiller) [inline src/core/lib/channel/http_client_filter.c]: This needs to pass the original op down the stack

```c
grpc_call_next_op(..., calld->saved_send_message_op);
```

File: src/core/lib/channel/http_client_filter.h (hunk 4 above)
Comment (ctiller) [inline src/core/lib/channel/http_client_filter.h]: the kMaxPayloadSizeForGet needn't be exposed though

File: src/core/lib/channel/http_client_filter.c (hunk 6 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: OK

File: src/core/lib/channel/http_client_filter.c (hunk 8 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: grpc_call_next_op is called in got_slice closure.

File: src/core/lib/channel/http_client_filter.h (hunk 4 above)
Comment (ctiller) [inline src/core/lib/channel/http_client_filter.h]: Ping on this one.

**Comment by ctiller** [comment]
LGTM

**Pull request merged**